compiler/rustc_codegen_llvm/src/type_of.rs

```diff
@@ -311,7 +311,7 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyAndLayout<'tcx> {
             }
             // only wide pointer boxes are handled as pointers
             // thin pointer boxes with scalar allocators are handled by the general logic below
-            ty::Adt(def, args) if def.is_box() && cx.layout_of(args.type_at(1)).is_zst() => {
+            ty::Adt(def, args) if def.is_box() && cx.layout_of(args.type_at(1)).is_1zst() => {
```

Comment:
I think this is also a bugfix -- a `Box` with an allocator that is ZST but highly aligned (more aligned than a pointer) cannot just be treated like a raw pointer.